arch/xtensa/src/esp32/esp32_spi.c

```diff
@@ -0,0 +1,1076 @@
+  uint32_t clk_insig;         /* SPI CLK input signal index */
+  uint32_t clk_outsig;        /* SPI CLK output signal index */
+};
+
+struct esp32_spi_priv_s
+{
+  /* Externally visible part of the SPI interface */
+
+  struct spi_dev_s spi_dev;
+
+  /* Port configuration */
+
+  const struct esp32_spi_config_s *config;
+
+  int refs;                    /* Referernce count */
+
+  /* Held while chip is selected for mutual exclusion */
+
+  sem_t            exclsem;
+
+  uint32_t         frequency;   /* Requested clock frequency */
+  uint32_t         actual;      /* Actual clock frequency */
+
+  enum spi_mode_e  mode;        /* Actual SPI hardware mode */
+
+  /* Actual SPI send/receive bits once transmission */
+
+  uint8_t          nbits;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int esp32_spi_lock(FAR struct spi_dev_s *dev, bool lock);
+static void esp32_spi_select(FAR struct spi_dev_s *dev,
+                             uint32_t devid,
+                             bool selected);
+static uint32_t esp32_spi_setfrequency(FAR struct spi_dev_s *dev,
+                                       uint32_t frequency);
+static void esp32_spi_setmode(FAR struct spi_dev_s *dev,
+                              enum spi_mode_e mode);
+static void esp32_spi_setbits(FAR struct spi_dev_s *dev, int nbits);
+#ifdef CONFIG_SPI_HWFEATURES
+static int esp32_spi_hwfeatures(FAR struct spi_dev_s *dev,
+                                spi_hwfeatures_t features);
+#endif
+#ifdef CONFIG_SPI_CMDDATA
+static int esp32_spi_cmddata(FAR struct spi_dev_s *dev,
+                             uint32_t devid,
+                             bool cmd);
+#endif
+static uint32_t esp32_spi_send(FAR struct spi_dev_s *dev, uint32_t wd);
+static void esp32_spi_exchange(FAR struct spi_dev_s *dev,
+                               FAR const void *txbuffer,
+                               FAR void *rxbuffer, size_t nwords);
+#ifndef CONFIG_SPI_EXCHANGE
+static void esp32_spi_sndblock(FAR struct spi_dev_s *dev,
+                               FAR const void *txbuffer,
+                               size_t nwords);
+static void esp32_spi_recvblock(FAR struct spi_dev_s *dev,
+                                FAR void *rxbuffer,
+                                size_t nwords);
+#endif
+#ifdef CONFIG_SPI_TRIGGER
+static int esp32_spi_trigger(FAR struct spi_dev_s *dev);
+#endif
+static void esp32_spi_init(FAR struct spi_dev_s *dev);
+static void esp32_spi_deinit(FAR struct spi_dev_s *dev);
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+extern uint8_t esp32_spi2_status(FAR struct spi_dev_s *dev, uint32_t devid);
+extern uint8_t esp32_spi3_status(FAR struct spi_dev_s *dev, uint32_t devid);
```

Comment:
For other architectures, these are prototyped in the header file (arch/xtensa/src/esp32/esp32_spi.h in this case) and implemented in the board logic.

Reply:
Yes, I'm moving it to the header